### PR TITLE
Added noexcept to methods called by async functions

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

The frameworks internal async functions assume that no exceptions happen during the call. By adding noexcept to methods they call the C++ runtime will enforce this requirement.

#### PR validation:

Framework tests pass.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
Does this PR address https://github.com/cms-sw/cmssw/issues/44881 fully? Was there anything suspicious being called by the "leaf `noexcept`" functions? (or is figuring out those part of the review?)

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
> Does this PR address https://github.com/cms-sw/cmssw/issues/44881 fully? 
I think so, except for the _services could throw exceptions_ which we spotted earlier.

> Was there anything suspicious being called by the "leaf noexcept" functions? (or is figuring out those part of the review?)

I didn't notice anything, but wasn't looking super closely.

File: FWCore/Framework/src/ProductResolvers.cc [hunk 1]
```diff
@@ -996,7 +990,7 @@ namespace edm {
                                                 bool skipCurrentProcess,
                                                 ServiceToken const& token,
                                                 SharedResourcesAcquirer* sra,
-                                                ModuleCallingContext const* mcc) const {
+                                                ModuleCallingContext const* mcc) const noexcept {
```
Comment (makortel) [inline FWCore/Framework/src/ProductResolvers.cc:993]: We should inspect `NoProcessProductResolver::tryPrefetchResolverAsync()`.

File: FWCore/Framework/src/ProductResolvers.cc [hunk 2]
```diff
@@ -1109,7 +1103,7 @@ namespace edm {
                                                           SharedResourcesAcquirer* sra,
                                                           ModuleCallingContext const* mcc,
                                                           ServiceToken token,
-                                                          oneapi::tbb::task_group* group) const {
+                                                          oneapi::tbb::task_group* group) const noexcept {
```
Comment (makortel) [inline FWCore/Framework/src/ProductResolvers.cc:1106]: We should inspect `Principal:: getProductResolverByIndex()` (probably just mark it as noexcept)

**Review by makortel — COMMENTED** [review]
Noted a few places we could continue `noexcept` percolation (in subsequent PR(s)).

**Comment by makortel** [comment]
+core

**Comment by perrotta** [comment]
+db
- Technical...

**Comment by rappoccio** [comment]
+1

**Pull request merged**